File: src/Infrastructure/Vpiska.Mongo/Entry.cs
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using MongoDB.Bson.Serialization;
using MongoDB.Bson.Serialization.Conventions;
using MongoDB.Driver;
using Vpiska.Domain;

namespace Vpiska.Mongo
{
    public static class Entry
    {
        public static void AddMongo(this IServiceCollection services, IConfigurationSection mongoSection)
        {
            var conventionPack = new ConventionPack
            {
                new CamelCaseElementNameConvention(),
                new ImmutableTypeClassMapConvention()
            };
            
            ConventionRegistry.Register("default", conventionPack, t => true);
            
            BsonClassMap.RegisterClassMap<User>(cm => 
            {
                cm.AutoMap();
                cm.MapIdMember(c => c.Id);
            });

            var client = new MongoClient(mongoSection["ConnectionString"]);
            services.AddSingleton(client);
            services.AddTransient<MongoUserRepository>();
        }
    }
}

File: src/Vpiska.Infrastructure/Vpiska.Infrastructure.Mongo/Entry.fs
namespace Vpiska.Infrastructure.Mongo

open System.Runtime.CompilerServices
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open MongoDB.Bson.Serialization
open MongoDB.Bson.Serialization.Conventions
open MongoDB.Driver
open Vpiska.Domain.User

type MongoSettings =
    { ConnectionString: string
      DatabaseName: string }

[<Extension>]
type Entry() =
    
    [<Extension>]
    static member AddMongo(services: IServiceCollection, mongoSection: IConfigurationSection) =
        let conventionPack = ConventionPack()
        CamelCaseElementNameConvention() |> conventionPack.Add
        ImmutableTypeClassMapConvention() |> conventionPack.Add
        ConventionRegistry.Register("default", conventionPack, fun t -> true)
        BsonClassMap.RegisterClassMap<User>(fun cm -> cm.AutoMap(); cm.MapIdMember(fun c -> c.Id) |> ignore) |> ignore
        let settings = { ConnectionString = mongoSection.["ConnectionString"]
                         DatabaseName = mongoSection.["DatabaseName"] }
        let client = MongoClient settings.ConnectionString
        services.AddSingleton(settings) |> ignore
        services.AddSingleton(client) |> ignore
    